DCC 5103 - i18n implementation

SidSyrus opened the pull request:
- angular-gettext dependency added to bower file
- new grunt-anuglat-gettext plugin introduced to package.json
- Documentation on Internationalization
- angular-gettext components added to karma
  -HTML, JavaScripts strings annotated
- new tasks for extracting and compiling translations introduced in Gruntfile

Commit:
- new angular-gettext plugin in bower.json
- new grunt-anuglat-gettext plugin in package.json
- Internationalization DOC file
- angular-gettext JS file added to karma
- String annoyed
- new task introduced in Gruntfile

Commit:
Edit to the Internationalization documentation, gitignore file updated

SidSyrus committed:
Rename

Renamed Internationalization.md to INTERNATIONALIZATION.md

Commit:
String.pot added

Commit:
Edit to the documentation

Commit:
Merge branch 'feature/DCC-5103-i18n-implementation' of https://github.com/icgc-dcc/dcc-portal into feature/DCC-5103-i18n-implementation

# Conflicts:
#	dcc-portal-ui/INTERNATIONALIZATION.md

Commit:
Documentation updated, more strings annotated , strings.pot file updated, new filter introduced for repositories to work with Internationalization of 'Experimental&nbsp;Strategy'

Commit:
Merge conflict
- dcc-portal-ui/app/scripts/common/views/footer.html
-dcc-portal-ui/app/scripts/software/js/software.js
-dcc-portal-ui/app/scripts/software/html/software.html

btiernay commented on dcc-portal-ui/INTERNATIONALIZATION.md:
```diff
@@ -139,5 +139,5 @@ language code. This code can be found in the `.po` file header.
 are listed on [Wikipedia](http://en.wikipedia.org/wiki/Gettext#See_also). The `.pot` file 
 generated by Grunt plugin is a standard file format for any translator. This file can be
-found in [translations](dcc-portal-ui/translations/) folder. Use any of the listed services or a translator 
+found in [translations](/translations/) folder. Use any of the listed services or a translator 
```
Can we add some links to the IDE you were using and the translation services?

SidSyrus replied:
The link (http://en.wikipedia.org/wiki/Gettext#See_also) to Wikipedia I mentioned has the services and IDE I was using.

Commit:
added translations.js to JSHint ignore list, increased maxlen for JS files.
annotated more strings
some JS file strings are now being templated using loadsh
added English PO file
App now sets English as default language when it runs

Commit:
Development MERGE

Commit:
Missing ')' added

Commit:
DOC updated, Comments removed, Karma error fixed

SidSyrus commented:
Deployed on Slot no. 21

Commit:
Work around for constants. new annoyed strings.

Commit:
Work arounds added to the documentations, attributes with interpolation updated, spelling mistakes

btiernay commented:
Are you sure it won't work? Why would it treat a variable name as a string? Did you try it?

Commit:
Doc updated, translations.js added to repo, tooltips with interpolation added, fix spelling mistake.

Commit:
Some missing text annotated

cheapsteak commented on dcc-portal-ui/INTERNATIONALIZATION.md:
```diff
@@ -0,0 +1,226 @@
+
+- Initial setup for angular-gettext has been done for Portal. You can just run `bower install` command from the root directory of **dcc-portal-ui** and it will install needed module to start developing. Below information is just for knowledge. 
+
+- Install the angular-gettext plugin
```
> Below information is just for knowledge 

I don't think we should keep these

Documentation should ideally be succinct, 
"This is what you need to do to use our thing", not "This is what we did to make this possible". 

Save this in a [gist](https://gist.github.com) for when you write the blog post on the process, but people reading these docs to figure out how to either setup their own portal, or how to use the translation service that's been setup, don't need to know this.

SidSyrus replied:
It would be nice to have. If they are setting up their own portal they can see how the internationalization is setup. In case they want to fiddle with it.

cheapsteak replied:
Hmm, our cost/benefit functions seem to greatly differ

We're essentially duplicating their documentation, I think we should at most link to it instead 

"To see how internationalization was setup, see angular-gettext's [Get Started page](https://angular-gettext.rocketeer.be/dev-guide/install/)". Although even that seems unnecessary since we're already linking to angular-gettext in the first sentence

We also don't have instructions on how to setup Angular, Karma, d3.

cheapsteak replied:
Installation instructions for things that have already been installed just seems incredibly out of scope

What is there that could be fiddled with?

cheapsteak commented on dcc-portal-ui/INTERNATIONALIZATION.md:
```diff
@@ -0,0 +1,226 @@
+Strings can be extracted using different build tools like Grunt, Gulp and webpack. However, Grunt
+is the only one officially supported by the angular-gettext community. The build tool will extract
+the strings in a `.pot` file: a standard gettext template that lists all the strings that should be
+translated. Grunt plugin is already configured for the portal. You will only need to run extract task `grunt extractText` and compile task `grunt compileText` once you have the desired translation. Below information is just for your knowledge. 
```
> Below information is just for your knowledge

same here

SidSyrus commented:
@btiernay Yes, I tried. It won't work because the extractor doesn't know its a variable. It only extracts the string surrounded by translate filter.

cheapsteak commented on dcc-portal-ui/app/scripts/repository/js/repository.js:
```diff
@@ -1131,4 +1132,11 @@
+  module.filter('addEntity', function(){
+    return function(string){
+      return string.split(' ').join('\u00A0');
+    };
```
When naming functions, filters, classes, think in terms of "would someone who only sees where this is used (without reading the function definition, or the documentation) be able to understand what it does? (or at least be able to make a decent guess)"

If you were to ask someone what they would assume a filter called `addEntity` does, I don't think they'd be able to guess that it  "replaces spaces with non breaking white space characters". 

Also not a fan of creating functions that can only do very specific things, unless it's used so frequently where it's worth it to create a partially bound variant of the function

In this instance, a `replace` filter would be appropriate

Usage would be something like

``` html
{{'Experimental Strategy' | translate | replace: ' ':'\u00A0'}}
```

This makes it very clear what is happening, and the filter is also flexible enough to be useful in other places

In our case, since we already have lodash, we should be able to do this:

``` js
module.filter('replace', function(){
  return _.replace;
});
```

SidSyrus replied:
`_.replace` is available from lodash version 4.0.0 , while we are using 3.10.1. I have updated the code to use .replace strings JS function.

cheapsteak replied:
In that case

``` js

function replace (input, pattern, replacement) {
  return input.replace(new RegExp(pattern, 'g'), replacement)
}

module.filter('replace', function(){
  return replace;
});
```

Commit:
Code Review changes. Moved filter to common.js

cheapsteak commented on dcc-portal-ui/app/scripts/common/js/common.js:
```diff
@@ -379,4 +379,11 @@
+  // This is a workaroud required for Internationalization of 'Experimental&nbsp;Strategy'
+  module.filter('replace', function(){
+    return function(string, oldValue, newValue){
+      return string.replace(oldValue, newValue);
```
very close!👍 
unfortunately in js this will only replace the first occurrence
need a regex, see

``` js
function replace (input, pattern, replacement) {
  return input.replace(new RegExp(pattern, 'g'), replacement)
}
```

Commit:
filter update

btiernay commented:
Please address merge conflicts. Thanks

Commit:
Merge branch 'develop' into feature/DCC-5103-i18n-implementation

# Conflicts:
#	dcc-portal-ui/app/scripts/analysis/views/analysis.analysis.html

Commit:
Merge branch 'develop' into feature/DCC-5103-i18n-implementation

cheapsteak commented on dcc-portal-ui/app/scripts/repository/views/repository.external.html:
```diff
@@ -181,20 +183,20 @@ <h1 data-ui-scrollfix="79" class="sticky-scrollfix">
+                <pie items='ExternalRepoController.facetCharts.dataType' heading="{{'Data Type' | translate}}" label="{{'Files' | translate}}" type="pie" width="160" group-percent="3"
                    height="150"></pie>
-                <pie items='ExternalRepoController.facetCharts.experimentalStrategy' heading="Experimental&nbsp;Strategy" label="Files" type="pie" width="160" group-percent="3"
+                <pie items='ExternalRepoController.facetCharts.experimentalStrategy' heading="{{'Experimental Strategy' | translate | replace: ' ':'&nbsp;'}}" label="{{'Files' | translate}}" type="pie" width="160" group-percent="3"
```
Let's add this filter `| replace: ' ':'&nbsp;'` to "Primary Site"
